spec/helpers/http_mock/asserts.lua

```diff
@@ -0,0 +1,153 @@
+local pcall = pcall
+local error = error
+
+local http_mock = {}
+
+local build_in_checks = {}
+
+local eventually_MT = {}
+eventually_MT.__index = eventually_MT
+
+local step_time = 0.01
+
+-- example for a check function
+-- local function(session, status)
+--   -- must throw error if the assertion is not true
+--   -- instead of return false
+--   assert.same(session.resp.status, status)
+--   -- return a string to tell what condition is satisfied
+--   -- so we can construct an error message for reverse assertion
+--   -- in this case it would be "we don't expect that: has a response with status 200"
+--   return "has a response with status " .. status
+-- end
+
+local function eventually_has(check, mock, ...)
+  local time = 0
+  local ok, err
+  while time < mock.eventually_timeout do
+    local records = mock:get_record()
+    for _, record in ipairs(records) do
+    -- use pcall so the user may use lua assert like assert.same
+      ok, err = pcall(check, record, ...)
+      if ok then
+        return true
+      end
+    end
+    time = time + step_time
+  end
+
+  error(err or "assertion fail", 2)
+end
+
+-- wait until timeout to check if the assertion is true for all records
+local function eventually_all(check, mock, ...)
+  local time = 0
+  local ok, err
+  while time < mock.eventually_timeout do
+    local records = mock:get_record()
+    for _, record in ipairs(records) do
+      ok, err = pcall(check, record, ...)
+      if not ok then
+        error(err or "assertion fail", 2)
+      end
+    end
+    time = time + 0.01
+  end
+
+  return true
+end
+
+-- a session is a request/response pair
+function build_in_checks.session_satisfy(session, f)
+  return f(session) or "session satisfy"
+end
+
+function build_in_checks.request_satisfy(session, f)
+  return f(session.req) or "request satisfy"
+end
+
+function build_in_checks.response_satisfy(session, f)
+  return f(session.resp) or "response satisfy"
+end
+
+function build_in_checks.error_satisfy(session, f)
+  return f(session.err) or "error satisfy"
+end
+
+function build_in_checks.error(session)
+  assert(session.err, "has no error")
+  return "has error"
+end
+
+local function register_assert(name, impl)
+  eventually_MT["has_" .. name] = function(self, ...)
+    return eventually_has(impl, self.__mock, ...)
+  end
+
+  eventually_MT["all_" .. name] = function(self, ...)
+    return eventually_all(impl, self.__mock, ...)
+  end
+
+  local function reverse_impl(session, ...)
+    local ok, err = pcall(impl, session, ...)
+    if ok then
+      error("we don't expect that: " .. (name or err), 2)
+    end
+    return true
+  end
+
+  eventually_MT["has_no_" .. name] = function(self, ...)
+    return eventually_all(reverse_impl, self.__mock, ...)
+  end
+
+  eventually_MT["not_all_" .. name] = function(self, ...)
+    return eventually_has(reverse_impl, self.__mock, ...)
+  end
+end
+
+for name, impl in pairs(build_in_checks) do
+  register_assert(name, impl)
+end
```

Comment:
This looks cool and would probably be something to add to `helpers.lua` or live in a separate module.

Reply:
We have this: https://github.com/Kong/kong/blob/458fd9f2ee884c215a661133516462d296363efa/spec/helpers/wait.lua#L453

Actually, the design is inspired by `assert.eventually`.

This logic is coupled with `self.__mock` and `get_all_records()`. Maybe we will find a way to decouple them.